File: Project-Rails/app/controllers/concerns/applications.rb
module Applications
  extend ActiveSupport::Concern

  # XXX: Javascript: 1) make sure at least one animal has been chosen
  #                  2) verify "user_id", "age", "phone", "email", "postal_code", "income" field
  #                  3) if user exists, set all fields' values to the existing values
  #                  4) display request's response (@status)
  #                  5) redirect page and update @status
  # get 'applications/adopt' => 'petshelter#app_adopt'
  def app_adopt
    prepareBaseContext
    @animals = Animal.where('stage = ?', true)
    @files = Array.new
    for a in @animals
      @files.push((AnimalFile.find_by animal: a))
    end
    @people = People.take!
    if request.method == 'POST'
      @status = 'Adoption request(s) failed'
      user = People.find_by(user_id: request.POST['user_id'])
      if user == nil
        user = People.new
        user.user_id = request.POST['user_id']
        @status = 'Adoption request(s) succeeded'
      end
      user.user_type = 'USER'
      user.name = request.POST['name']
      user.gender = request.POST['gender'].upcase
      user.age = request.POST['age']
      user.phone = request.POST['phone']
      user.email = request.POST['email']
      user.address = request.POST['address']
      user.city = request.POST['city']
      user.province = request.POST['province']
      user.postal_code = request.POST['postal_code']
      user.education = request.POST['education']
      user.occupation = request.POST['occupation']
      user.income = request.POST['income']
      user.special_skills = request.POST['special_skills']
      user.save
      for animal in @animals
        if request.POST[animal.animal_id] != nil and Adoption.find_by(people: user, animal: animal) == nil
          adoption = Adoption.new
          adoption.people = user
          adoption.animal = animal
          adoption.apply_date = Time.now.strftime('%Y-%m-%d')
          adoption.adoption_date = nil
          adoption.accepted = false
          adoption.save
          @status = 'Adoption request(s) succeeded'
        end
      end
      redirect_to action: 'app_adopt'
      return
    end # if request.method == 'POST'
    render 'petshelter/applications/adopt'
  end # def app_adopt

  # XXX: Only file uploading part is ready. Record updating need to be done
  #      Besides, you may want rename the file instead of using it's original name
  # XXX: How to allow uploading multiple files???
  # XXX: Javascript: 1) make sure at least one animal has been surrendered
  #                  2) verify "user_id", "age", "phone", "email", "postal_code", "income" field
  #                  3) if user exists, set all fields' values to the existing values
  #                  4) display request's response (@status)
  #                  5) redirect page and update @status
  # get 'applications/surrender' => 'petshelter#app_surrender'
  def app_surrender
    prepareBaseContext
    @animal = Animal.take!
    @people = People.take!
    if request.method == 'POST'
      uploaded_io = params[:photo]
      if uploaded_io != nil
        folder = ''
        if ['.jpg', '.png', '.gif', '.bmp'].include? File.extname(uploaded_io.original_filename)
          folder = 'images'
        elsif ['.wav', '.mp3', '.ogg'].include? File.extname(uploaded_io.original_filename)
          folder = 'audios'
        elsif ['.mp4', '.ogg', '.webm'].include? File.extname(uploaded_io.original_filename)
          folder = 'videos'
        end
        File.open(Rails.root.join('public', folder, uploaded_io.original_filename), 'wb') do |file|
          file.write(uploaded_io.read)
        end
      end
    end
    render 'petshelter/applications/surrender'
  end # def app_surrender

  # get 'applications/donate' => 'petshelter#app_donates'
  def app_donates
    prepareBaseContext
    @animals = Animal.where('stage = ?', true)
    @files = Array.new
    for a in @animals
      @files.push((AnimalFile.find_by animal: a))
    end
    @people = People.take!
    render 'petshelter/applications/donates'
  end # def app_donates

  # get 'applications/donate/:donate_type' => 'petshelter#app_donate'
  def app_donate
    prepareBaseContext
    @donate_type = params[:donate_type]
    @animals = Animal.where('stage = ?', true)
    @files = Array.new
    for a in @animals
      @files.push((AnimalFile.find_by animal: a))
    end
    @people = People.take!
    render 'petshelter/applications/donate'
  end # def app_donate

  # get 'applications/volunteer' => 'petshelter#app_volunteer'
  def app_volunteer
    prepareBaseContext
    @shelters = Shelter.all
    @people = People.take!
    render 'petshelter/applications/volunteer'
  end # def app_volunteer

end # ﻿module Applications

File: Project-Django/PetShelter/views/applications.py
from django.shortcuts import render, redirect, get_object_or_404
from PetShelter.models import People, Shelter, ShelterStaff, Animal, AnimalFile,\
                              Receipt, Adoption, Surrender, Donation
from PetShelter.methods import *
from PetShelter.forms import UploadFileForm
from django.contrib.auth.models import User
from django.contrib.auth import authenticate
from django.contrib.auth import login as auth_login
from django.contrib.auth import logout as auth_logout
from django.contrib.auth.decorators import login_required
from django.core.exceptions import ObjectDoesNotExist
import datetime

# url(r'^applications/adopt/$', views.app_adopt, name='app_adopt'),
def app_adopt(request):
  animals = Animal.objects.filter(stage=True)
  files = []
  for a in animals:
    files.append(AnimalFile.objects.filter(animal=a, file_type='PHOTO')[0])
  context = {
    'animals': animals,
    'files': files,
    'people': People.objects.all()[0],
  } # context = {...}
  if request.method == 'POST':
    user_id=request.POST['user_id']
    name = request.POST['name']
    gender = request.POST['gender'].upper()
    age = 0 if request.POST['age'] == '' else int(request.POST['age'])
    phone = request.POST['phone']
    email = request.POST['email']
    address = request.POST['address']
    province = request.POST['province']
    postal_code = request.POST['postal_code']
    education = request.POST['education']
    occupation = request.POST['occupation']
    income = 0.0 if request.POST['income'] == '' else float(request.POST['income'])
    special_skills = request.POST['special_skills']
    try:
      u = People.objects.get(user_id=user_id)
      u.name = name
      u.gender = gender
      u.age = age
      u.phone = phone
      u.email = email
      u.address = address
      u.province = province
      u.postal_code = postal_code
      u.education = education
      u.occupation = occupation
      u.income = income
      u.special_skills = special_skills
    except ObjectDoesNotExist:
      u = People(user_id=user_id, user_type='USER', name=name, gender=gender, age=age, phone=phone, \
                 email=email, address=address, province=province, postal_code=postal_code, \
                 education=education, occupation=occupation, income=income, special_skills=special_skills)
    u.save()
    animals = request.POST.getlist('adoptions')
    now = datetime.datetime.now()
    success = False
    msg = ''
    for a in animals:
      animal = Animal.objects.get(animal_id=a)
      if len(Adoption.objects.filter(animal=animal, accepted=True)) == 0:
        Adoption(application_date=now, adopter=u, animal=animal).save()
        success = True
        msg += '%s\n' % str(animal)
    context['status'] = 'Adoption(s) requested: \n' + msg if success else 'Fail'
  return render(request, 'PetShelter/applications/adopt.html', getBaseContext(request, context))
# def app_adopt(request)

# url(r'^applications/surrender/$', views.app_surrender, name='app_surrender'),
def app_surrender(request):
  context = {
    'animal': Animal.objects.all()[0],
    'people': People.objects.all()[0],
    'form': UploadFileForm,
  } # context = {...}
  msg = ''
  if request.method == 'POST':
    if len(request.FILES.keys()) != 0:
      form = UploadFileForm(request.POST, request.FILES)
      if form.is_valid():
        handle_uploaded_file(request.FILES['file'], str(datetime.datetime.now()))
        return redirect('app_surrender')
      # if form.is_valid()
    # if len(request.FILES.keys()) != 0
  # if request.method == 'POST'
  success = True
  context['status'] = 'Adoption(s) requested: \n' + msg if success else 'Fail'
  return render(request, 'PetShelter/applications/surrender.html', getBaseContext(request, context))
# def app_surrender(request)

# url(r'^applications/donate/$', views.app_donates, name='app_donates'),
def app_donates(request):
  animals = Animal.objects.filter(stage=True)
  files = []
  for a in animals:
    files.append(AnimalFile.objects.filter(animal=a, file_type='PHOTO')[0])
  context = {
    'animals': animals,
    'files': files,
    'people': People.objects.all()[0],
  } # context = {...}
  return render(request, 'PetShelter/applications/donates.html', getBaseContext(request, context))
# def app_donates(request)

# url(r'^applications/donate/(?P<donate_type>[^\s/]+)/$', views.app_donate, name='app_donate'),
def app_donate(request, donate_type):
  animals = Animal.objects.filter(stage=True)
  files = []
  for a in animals:
    files.append(AnimalFile.objects.filter(animal=a, file_type='PHOTO')[0])
  context = {
    'animals': animals,
    'files': files,
    'people': People.objects.all()[0],
    'donate_type': donate_type.lower(),
  } # context = {...}
  return render(request, 'PetShelter/applications/donate.html', getBaseContext(request, context))
# def app_donate(request)

# url(r'^applications/volunteer/$', views.app_volunteer, name='app_volunteer'),
def app_volunteer(request):
  context = {
    'shelters': Shelter.objects.all(),
    'people': People.objects.all()[0],
  } # context = {...}
  return render(request, 'PetShelter/applications/volunteer.html', getBaseContext(request, context))
# def app_volunteer(request)

